src/stepFunctions/commands/visualizeStateMachine.ts

```diff
@@ -29,257 +29,316 @@ export interface messageObject {
-     * Ensure tests are written for this use case as well.
-     */
-    const activeTextEditor: vscode.TextEditor | undefined = vscode.window.activeTextEditor
-
-    let documentUri: vscode.Uri
-    let textDocument: vscode.TextDocument
-
-    if (activeTextEditor) {
-        documentUri = activeTextEditor.document.uri
-        textDocument = activeTextEditor.document
-    } else {
-        logger.error('Could not get active text editor for state machine render.')
-        throw new Error('Could not get active text editor for state machine render.')
-    }
+            return existingVisualization.getWebviewPanel()
+        }
 
-    try {
-        await cache.updateCache(globalStorage)
+        // Existing visualization does not exist, construct new visualization
+        try {
+            await cache.updateCache(globalStorage)
 
-        return setupWebviewPanel(documentUri, textDocument)
-    } catch (err) {
-        vscode.window.showInformationMessage(
-            localize(
-                'AWS.stepfunctions.visualisation.errors.rendering',
-                'There was an error rendering State Machine Graph, check logs for details.'
+            return this.createNewVisualization(textDocument)
+        } catch (err) {
+            vscode.window.showInformationMessage(
+                localize(
+                    'AWS.stepfunctions.visualisation.errors.rendering',
+                    'There was an error rendering State Machine Graph, check logs for details.'
+                )
             )
-        )
 
-        logger.debug('Unable to setup webview panel.')
-        logger.error(err as Error)
+            logger.debug('Unable to setup webview panel.')
+            logger.error(err as Error)
+        }
+
+        return
     }
 
-    return
-}
+    public removeClosedPanel(visToDelete: aslVisualization): void {
+        this.managedVisualizations.delete(visToDelete)
+    }
 
-async function setupWebviewPanel(
-    documentUri: vscode.Uri,
-    textDocument: vscode.TextDocument
-): Promise<vscode.WebviewPanel> {
-    const logger: Logger = getLogger()
-    let isPanelDisposed = false
-
-    // Create and show panel
-    const panel = vscode.window.createWebviewPanel(
-        'stateMachineVisualization',
-        makeWebviewTitle(documentUri),
-        {
-            preserveFocus: true,
-            viewColumn: vscode.ViewColumn.Beside,
-        },
-        {
-            enableScripts: true,
-            localResourceRoots: [
-                ext.visualizationResourcePaths.localWebviewScriptsPath,
-                ext.visualizationResourcePaths.visualizationLibraryCachePath,
-                ext.visualizationResourcePaths.stateMachineCustomThemePath,
-            ],
-            retainContextWhenHidden: true,
-        }
-    )
+    private createNewVisualization(textDocument: vscode.TextDocument): vscode.WebviewPanel {
+        const newVisualization = new aslVisualization(textDocument, this)
```

Comment:
There's some unfortunate coupling going on here... aslVisualizationManager  creates instances of aslVisualization but aslVisualization deletes itself from aslVisualizationManager. 

A clearer separation of concerns could improve the maintainability and testability of each class. Currently, I think this code is only covered by integration tests, it would be ideal for some unit/component tests to cover portions of this file's code where appropriate.